lib/src/main/java/com/futurewei/alcor/common/db/ignite/IgniteClientDbCache.java

```diff
@@ -37,66 +44,113 @@
-            logger.log(Level.INFO, "Cache " + name + " AtomicityMode is " + this.cache.getConfiguration().getAtomicityMode());
+            if (!checkedForSqlFields) {
+                checkedForSqlFields = true;
+                extractSqlFields(className);
+                if (sqlFields != null && sqlFields.size() != 0) {
+                    ClientCacheConfiguration clientCacheConfig = new ClientCacheConfiguration();
+                    clientCacheConfig.setName(CommonUtil.getSqlNameFromCacheName(name));
+                    this.cache = getOrCreateIndexedCache(igniteClient, className, clientCacheConfig, null);
+                    if (this.cache == null) {
+                        logger.log(Level.WARNING, "Create cache for client " + className + " with index failed, falling back");
+                    }
+                }
+            }
+            if (this.cache == null)
+                this.cache = igniteClient.getOrCreateCache(className);
+
         } catch (ClientException e) {
-            logger.log(Level.WARNING, "Create cache for client " + name + " failed:" + e.getMessage());
+            logger.log(Level.WARNING, "Create cache for client " + className + " failed: " + e.getMessage());
         }
 
-        Assert.notNull(this.cache, "Create cache for client " + name + "failed");
+        Assert.notNull(this.cache, "Create cache for client " + className + " failed");
+        logger.log(Level.INFO, "Cache " + className + " AtomicityMode is " + this.cache.getConfiguration().getAtomicityMode());
         this.transaction = new IgniteClientTransaction(igniteClient);
     }
 
-    public IgniteClientDbCache(IgniteClient igniteClient, CacheConfiguration cacheConfig) {
+    public IgniteClientDbCache(IgniteClient igniteClient, Class<?> v, CacheConfiguration cacheConfig) {
         try {
+            String className = v.getName();
             ClientCacheConfiguration clientCacheConfig = new ClientCacheConfiguration();
-            clientCacheConfig.setName(cacheConfig.getName());
+            clientCacheConfig.setName(CommonUtil.getSqlNameFromCacheName(cacheConfig.getName()));
             clientCacheConfig.setAtomicityMode(cacheConfig.getAtomicityMode());
             logger.log(Level.INFO, "Getting or creating cache " + clientCacheConfig.getName() + " AtomicityMode is " + clientCacheConfig.getAtomicityMode());
-            this.cache = igniteClient.getOrCreateCache(clientCacheConfig);
-            logger.log(Level.INFO, "Retrieved cache " +  this.cache.getConfiguration().getName() + " AtomicityMode is " + this.cache.getConfiguration().getAtomicityMode());
+            extractSqlFields(className);
+            if (sqlFields != null && sqlFields.size() != 0) {
+                this.cache = getOrCreateIndexedCache(igniteClient, className, clientCacheConfig, null);
+                if (this.cache == null) {
+                    logger.log(Level.WARNING, "Create cache for client " + className + " with index failed, falling back");
```

Comment:
@pkommoju How to handle falling back if `this.cache=null` ? just use `igniteClient.getOrCreateCache(clientCacheConfig)` instead here?

Reply:
I am logging it as a warning message here. Few lines below, if this.cache is null, we call igniteClient.getOrCreateCache(clientCacheConfig). This call could also be because the the cache is not setup for SQLFieldsQuery in addition to handling the case where it is setup but something went wrong. Didn't want to have the call in two places.